### PR TITLE
fix #7045 adding @return $this on the docs

### LINKED ISSUE #7045: static vs. $this in /** @return */ docblock tag
I got very confused about the meaning of `/** @return static */` in the class reference. A long and confusing discussion ensued on the #yii channel. In the end we agreed that the docs are inconsistent and it is probably for the best to adopt the convention specified by [phpDocumentor](http://www.phpdoc.org/docs/latest/guides/types.html), specifically:

> The PHPDoc Standard also describes several keywords that are not native to PHP but are found to be often used or are representations of situations that are convenient to describe.

>    - **static** An object of the class where this value was consumed, if inherited it will represent the child class. (see late static binding in the PHP manual).
>    - **$this** This exact object instance, usually used to denote a fluent interface.

[For those unfamiliar with these terms, *late static binding* is using the `static` keyword other than for declaring a static member. *Fluent interfaces* means chaining methods, e.g. `$query->from()->limit()->where()`.]

It turns out that PhpStorm understands things this way too. You can try this for yourself. If you let PhpStorm automatically write the docblocks for the following two methods, this is the result

```php
<?php

class Foo
{
    /**
     * @return $this
     */
    public function bar()
    {
        return $this;
    }

    /**
     * @return static
     */
    public function baz()
    {
        return new static();
    }
}
```

And, tbh, I just find it hard to get my head around the idea of $this being static.

I propose:

1. Yii 2.0 adopt the convention as specified on www.phpdoc.org and quoted above
2. Reiterate it in the style guide wiki
3. Consider adding to the front page of the (highly formal) Class Reference a section called "Document conventions" where various things like this can be explained to readers.
4. Update docblocks in the framework code to conform to this convention

### PR DESCRIPTION
**Pull request by Faryshta**
edit: i think i am done and this pr is ready to merge

I submit the PR early so that you can check if this is what you need. so far the framework folder has been completed, check to see if its what you want and need, then i will change the other folders

### CODE REVIEW

**Comment by Faryshta** [comment]
why are all the PR about documentation falling the travis test?

**Comment by samdark** [comment]
It's broken now for everything.

**Comment by samdark** [comment]
https://github.com/yiisoft/yii2/issues/7045

File: extensions/authclient/BaseOAuth.php [hunk 1]
```diff
@@ -397,7 +397,7 @@ protected function composeUrl($url, array $params = [])
     /**
      * Saves token as persistent state.
      * @param OAuthToken $token auth token
-     * @return static self reference.
+     * @return $this
```
Comment (cebe) [inline extensions/authclient/BaseOAuth.php]: the comment should stay: `$this self reference.`
Also `self reference` could be improved to say a bit more.

Comment (Faryshta) [inline extensions/authclient/BaseOAuth.php]: don't you think thats redundant? $this is a self reference by definition.

besides as you can see on this pr, there was no consistency about what comment to add.

Comment (cebe) [inline extensions/authclient/BaseOAuth.php]: the return value should have a comment so one can read what is returned even if you do not know what `return $this` means.

Comment (Faryshta) [inline extensions/authclient/BaseOAuth.php]: reach an agreement about what should be that comment with the rest of the devs and i will edit it.

Comment (Faryshta) [inline extensions/authclient/BaseOAuth.php]: I was thinking. If I put

    @return Object

The api doc will change it to a link pointing to the documentation for object.

Can we do the same for `$this`?

**Comment by Faryshta** [comment]
@cebe any update on this?

**Comment by cwt137** [comment]
Putting 
```
@return $this
```
is wrong. You should put the the class name it is actually returning. So, for example, if a method inside DynamicModel returns $this, then the PHPdoc should say:

```
@return DynamicModel
```
This is the proper way. It does a few a things: 1) allows linking as @Faryshta mentioned above. 2) Allows the auto completion in a few IDEs to work, Putting the parent class Object is not correct because that is not the class being returned.

**Comment by dynasource** [comment]
exacly:

http://phpdoc.org/docs/latest/references/phpdoc/tags/return.html
Syntax
@return [Type] [<description>]

For types, see http://phpdoc.org/docs/latest/references/phpdoc/types.html

**Comment by Faryshta** [comment]
@samdark any update on this?

**Comment by samdark** [comment]
- [ ] Verify it works in PhpStorm.
- [ ] Verify it works in Netbeans.

**Comment by samdark** [comment]
https://github.com/yiisoft/yii2/commit/67658cf57d8bdda5c40d9a55894918075c163842

**Comment by samdark** [comment]
https://github.com/yiisoft/yii2-bootstrap/commit/25b717a43c16be1daa7d307ff601199f047927f8

**Comment by samdark** [comment]
https://github.com/yiisoft/yii2-gii/commit/c10afb648124fb27bebb803645a2af91d8f6651a

**Comment by samdark** [comment]
https://github.com/yiisoft/yii2-authclient/commit/c25b04a34ab1b9ae4f2f7a35aeee38033ad3654d

**Comment by samdark** [comment]
https://github.com/yiisoft/yii2-elasticsearch/commit/fe596835c52dc62278b409a9188c80cf45a39155

**Comment by samdark** [comment]
https://github.com/yiisoft/yii2-mongodb/commit/3f48f13dca6e59d025c00eb5a2de8e48481a355c

**Comment by samdark** [comment]
https://github.com/yiisoft/yii2-sphinx/commit/074e963e95478893fdefc33dcdd604a0eb3c1336

**Comment by samdark** [comment]
Thanks for the pointer.

**Pull request closed without merging**